runtime/src/iree/base/tree.c

```diff
@@ -0,0 +1,549 @@
+  } else if (x == x->parent->left) {
+    x->parent->left = y;
+  } else {
+    x->parent->right = y;
+  }
+  y->left = x;
+  x->parent = y;
+}
+
+static void iree_tree_rotate_right(iree_tree_t* tree, iree_tree_node_t* x) {
+  iree_tree_node_t* y = x->left;
+  x->left = y->right;
+  if (y->right != &tree->nil) {
+    y->right->parent = x;
+  }
+  y->parent = x->parent;
+  if (x->parent == NULL) {
+    tree->root = y;
+  } else if (x == x->parent->right) {
+    x->parent->right = y;
+  } else {
+    x->parent->left = y;
+  }
+  y->right = x;
+  x->parent = y;
+}
+
+static iree_status_t iree_tree_insert_internal(iree_tree_t* tree,
+                                               iree_host_size_t key,
+                                               iree_tree_node_t* t) {
+  t->left = &tree->nil;
+  t->right = &tree->nil;
+  t->key = key;
+  t->red = true;  // red
+  t->parent = NULL;
+
+  iree_tree_node_t* x = tree->root;
+  iree_tree_node_t* y = NULL;
+  while (x != &tree->nil) {
+    y = x;
+    if (t->key < x->key) {
+      x = x->left;
+    } else if (t->key > x->key) {
+      x = x->right;
+    } else {
+      return iree_make_status(IREE_STATUS_ALREADY_EXISTS,
+                              "Trying to insert a duplicate key");
+    }
+  }
+  t->parent = y;
+
+  if (!y) {
+    tree->root = t;
+  } else if (t->key < y->key) {
+    y->left = t;
+  } else {
+    y->right = t;
+  }
+
+  if (t->parent == NULL) {
+    t->red = false;
+    return iree_ok_status();
+  }
+
+  if (t->parent == tree->root) {
+    return iree_ok_status();
+  }
+
+  while (t->parent->red) {
+    if (t->parent == t->parent->parent->right) {
+      iree_tree_node_t* uncle = t->parent->parent->left;
+      if (uncle->red) {
+        uncle->red = false;
+        t->parent->red = false;
+        t->parent->parent->red = true;
+        t = t->parent->parent;
+      } else {
+        if (t == t->parent->left) {
+          t = t->parent;
+          iree_tree_rotate_right(tree, t);
+        }
+        t->parent->red = false;
+        t->parent->parent->red = true;
+        iree_tree_rotate_left(tree, t->parent->parent);
+      }
+    } else {
+      iree_tree_node_t* uncle = t->parent->parent->right;
+      if (uncle && uncle->red) {
+        uncle->red = false;
+        t->parent->red = false;
+        t->parent->parent->red = true;
+        t = t->parent->parent;
+      } else {
+        if (t == t->parent->right) {
+          t = t->parent;
+          iree_tree_rotate_left(tree, t);
+        }
+        t->parent->red = false;
+        t->parent->parent->red = true;
+        iree_tree_rotate_right(tree, t->parent->parent);
+      }
+    }
+    if (t == tree->root) {
+      break;
+    }
+  }
+  tree->root->red = false;
+
+  return iree_ok_status();
+}
+
+iree_status_t iree_tree_insert(iree_tree_t* tree, iree_host_size_t key,
+                               iree_tree_node_t** out_data) {
+  iree_tree_node_t* t;
+  iree_status_t status = iree_tree_allocate_node(tree, &t);
```

Comment:
IREE_RETURN_IF_ERROR on the first call or so long as no state is modified - after that you need status checks